test/AssetPool.test.js

```diff
@@ -975,4 +975,243 @@ describe("AssetPool", () => {
+          expect(
+            await collateralToken.balanceOf(newAssetPool.address)
+          ).to.equal(amountToUpgrade)
+        })
+
+        it("should transfer new underwriter tokens to the underwriter", async () => {
+          expect(
+            await newUnderwriterToken.balanceOf(underwriter1.address)
+          ).to.equal(amountToUpgrade)
+        })
+
+        it("should burn old underwriter tokens from the underwriter", async () => {
+          // 100 - 40 = 60
+          expect(
+            await underwriterToken.balanceOf(underwriter1.address)
+          ).to.equal(amountToDeposit.sub(amountToUpgrade))
+        })
+
+        it("should emit AssetPoolUpgraded event", async () => {
+          expect(tx)
+            .to.emit(assetPool, "AssetPoolUpgraded")
+            .withArgs(
+              underwriter1.address,
+              amountToUpgrade,
+              amountToUpgrade,
+              await lastBlockTime()
+            )
+        })
+      })
+      context("when there was a claim", () => {
+        const claim = to1e18(25)
+
+        beforeEach(async () => {
+          await assetPool
+            .connect(coveragePool)
+            .approveNewAssetPoolUpgrade(newAssetPool.address)
+          await assetPool.connect(underwriter1).deposit(amountToDeposit)
+          await assetPool
+            .connect(coveragePool)
+            .claim(coveragePool.address, claim)
+          await underwriterToken
+            .connect(underwriter1)
+            .approve(assetPool.address, amountToDeposit)
+
+          await assetPool
+            .connect(underwriter1)
+            .upgradeToNewAssetPool(amountToUpgrade)
+        })
+
+        it("should transfer collateral proportionally after the claim", async () => {
+          // underwriter 1 has 100/100 share of the pool
+          // claimed by the coverage pool: 25 which is 25%
+          // 40 * 75 / 100 = 30
+          expect(
+            await collateralToken.balanceOf(newAssetPool.address)
+          ).to.equal(to1e18(30))
+        })
+
+        it("should not change the underwriter's tokens to burn", async () => {
+          expect(
+            await underwriterToken.balanceOf(underwriter1.address)
+          ).to.equal(amountToDeposit.sub(amountToUpgrade))
+        })
+      })
+
+      context("when rewards were allocated", () => {
+        const allocatedReward = to1e18(70)
+
+        beforeEach(async () => {
+          await assetPool
+            .connect(coveragePool)
+            .approveNewAssetPoolUpgrade(newAssetPool.address)
+          await collateralToken
+            .connect(rewardManager)
+            .approve(rewardsPool.address, allocatedReward)
+          await rewardsPool.connect(rewardManager).topUpReward(allocatedReward)
+
+          await assetPool.connect(underwriter1).deposit(amountToDeposit)
+
+          await underwriterToken
+            .connect(underwriter1)
+            .approve(assetPool.address, amountToDeposit)
```

Comment:
Shouldn't `amountToUpgrade` be approved instead? Also, shouldn't we approve after one day passed. Should not do any difference technically but this is what will happen in the real world.

Reply:
18fed51

Reply:
> Also, shouldn't we approve after one day passed. Should not do any difference technically but this is what will happen in the real world.

This still needs to be addressed. For:
- when there was a claim
- when rewards were allocated

Reply:
Done in 089f488